pkg/proxy/iptables/proxier.go

```diff
@@ -1168,12 +1233,14 @@ func (proxier *Proxier) syncProxyRules() {
 		endpoints = endpoints[:0]
 		endpointChains = endpointChains[:0]
 		var endpointChain utiliptables.Chain
-		for _, ep := range proxier.endpointsMap[svcName] {
+		n := 0
+		for _, ep := range allEndpoints {
```

Comment:
I'm a bit confused.

This loop iterates all endpoints and builds up a list of endpoints and endpoint chains, but only counts the filtered ones in `n`.  We make a chain for each endpoint.

Below (L1266 and L1284) we loop for each endpoint chain. L1266 loop adds rules (affinity) to the chain.  L1284 detects that the chain is not valid for topology and aborts the loop. 

Why are we making a chain for endpoints that are not filtered?  If we know the filtered set, can't we just iterate that here, instead of "all" ?  That way we don't make chains that we don't need (keep iptables smaller).  If topology is disabled, we can arrange that filtered == all ?

Am I missing something?

Reply:
Nice catch! In the beginning, it was considered that the externalTrafficPolicy of the service might be Local, and the filteredEndpoints might not containts local endpoints, so we should iterate all endponints. But later we decide to make service topology mutually exclusive with externalTrafficPolicy=Local, so we can just iterate filtered endpoints if service topology enabled, and the `allEndpoints` is filtered endpoints.